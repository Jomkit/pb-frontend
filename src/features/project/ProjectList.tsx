import AddBtn from "../../components/AddBtn"
import { IProject } from "../../types"
import Project from "./Project"
import { useState } from "react"
import Popup from "../../components/Popup"
import NewProjectForm from "./NewProjectForm"


const ProjectList = ({projects}: {projects: IProject[]}) => {
  
  const [popupOpen, setPopupOpen] = useState(false);
  
  const addProject = () => {
    setPopupOpen(true);
  }

  const close = () => {
    setPopupOpen(false);
  }
  
  return (
    <div className="w-3/4 m-auto">
      <AddBtn handleClick={addProject} btnName="Project"/>
      {popupOpen && (
          <Popup close={close}>
            <NewProjectForm />
          </Popup>
        )}
      { projects.map((proj) => (
        <Project key={proj.id} project={proj} />
      ))}
    </div>
  )
}

export default ProjectList