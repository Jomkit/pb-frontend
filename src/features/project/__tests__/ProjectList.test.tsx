import { render, waitFor } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import ProjectList from "../ProjectList"
import userEvent from "@testing-library/user-event"

const testProjects = [
    {
        id: "1",
        name: "testProject 1",
        note: "testNote 1"
    },
    {
        id: "2",
        name: "testProject 2",
        note: "testNote 2"
    },
    {
        id: "3",
        name: "testProject 3",
        note: "testNote 3"
    }
]

it("should render without crashing", () => {
    render(
        <MemoryRouter>
            <ProjectList projects={testProjects} />
        </MemoryRouter>
    )
})

it("should match snapshot", () => {
    const { asFragment } = render(
        <MemoryRouter>
            <ProjectList projects={testProjects} />
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})

it("should open popup on click", async () => {
    const user = userEvent.setup();
    const { getByText, findByText } = render(
        <MemoryRouter>
            <ProjectList projects={testProjects} />
        </MemoryRouter>
    )
    const newProjectBtn = getByText("New Project");
    await user.click(newProjectBtn);
    
    await waitFor(async () => {
        const formTitle = await findByText("Create New Project");
        expect(formTitle).toBeInTheDocument();
    })
})

it("should close popup on click", async () => {
    const user = userEvent.setup();
    const { getByText, findByText } = render(
        <MemoryRouter>
            <ProjectList projects={testProjects} />
        </MemoryRouter>
    )
    const newProjectBtn = getByText("New Project");
    await user.click(newProjectBtn);
    
    const formTitle = await findByText("Create New Project");
    expect(formTitle).toBeInTheDocument();

    const closeButton = await findByText("âœ•");
    expect(closeButton).toBeInTheDocument();
    await user.click(closeButton);

    expect(formTitle).not.toBeInTheDocument();
})