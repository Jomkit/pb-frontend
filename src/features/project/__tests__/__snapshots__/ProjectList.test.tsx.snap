// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-3/4 m-auto"
  >
    <div
      class="text-start"
    >
      <button
        class="btn bg-green-400 border-solid border-2 border-black"
      >
        New Project
      </button>
    </div>
    <a
      href="/projects/1"
    >
      <div
        class="w-full mx-auto p-2 mb-2 border-solid border-2 border-black rounded-md bg-white"
        id="1"
      >
        <div
          class="header"
        >
          <h1
            class="text-3xl text-bold"
          >
            testProject 1
          </h1>
          <h3>
            Description: testNote 1
          </h3>
        </div>
      </div>
    </a>
    <a
      href="/projects/2"
    >
      <div
        class="w-full mx-auto p-2 mb-2 border-solid border-2 border-black rounded-md bg-white"
        id="2"
      >
        <div
          class="header"
        >
          <h1
            class="text-3xl text-bold"
          >
            testProject 2
          </h1>
          <h3>
            Description: testNote 2
          </h3>
        </div>
      </div>
    </a>
    <a
      href="/projects/3"
    >
      <div
        class="w-full mx-auto p-2 mb-2 border-solid border-2 border-black rounded-md bg-white"
        id="3"
      >
        <div
          class="header"
        >
          <h1
            class="text-3xl text-bold"
          >
            testProject 3
          </h1>
          <h3>
            Description: testNote 3
          </h3>
        </div>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
